Make sure to destroy PetscShellMatrix `_mat`

lindsayad opened the pull request:
With the change in #3914 to make the `_mat` data member the same as the `PetscMatrix(Base)` class, the manual destruction is now required. This refs the valgrind failures on idaholab/moose#28411

roystgnr commented:
Any chance we could accomplish this with a `WrappedPetsc<Mat>`?

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2393820/) on 6f04061 wanted to post the following:


## Coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>dec65e</th>
      <th colspan=4>#3932 6f0406</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>63.40%</td>
      <td>63.40%</td>
      <td>+0.01%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>71865</td>
      <td>71874</td>
      <td>+9</td>
      <td>3</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>41494</td>
      <td>41487</td>
      <td>-7</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/libmesh/docs/PRs/3932/coverage_diff)

[Full coverage report](https://mooseframework.inl.gov/libmesh/docs/PRs/3932/coverage)



This comment will be updated on new commits.

lindsayad commented:
That's what we used to do for this class and at one point while working on #3914 I switched `_mat` in the `PetscMatrix(Base)` to also be a `Wrapped` class. However, this didn't play nicely with some PETSc APIs for a "good" technical reason (unfortunately which I can't remember) so I changed both back to not be wrapped. This is something I could try exploring again in the future so at least I can publicly share why it doesn't work (or realize that it should work)

lindsayad commented:
It's coming back to me a little. There was definitely infinite recursion, I believe with `operator=` but I that bit I am fuzzy on
